### PR TITLE
spec: Access to cutting edge or undocumented REST APIs

### PR DESCRIPTION
**Pull request by mattheworiordan**
This PR provides access to raw REST APIs we decide to expose without having to roll these changes out to all client libraries.  For APIs we release progressively, this makes it a lot easier for developers to start using them until we decide to roll into the next official release, if at all.

### CODE REVIEW

**Comment by mattheworiordan** [comment]
@tcard FYI, this is a precursor to us providing support for push in all REST client libraries without exposing the full API which will likely fall into the 0.10 / 1.0 release.

cc @SimonWoolf @paddybyers

File: content/client-lib-development-guide/features.textile [hunk 1]
```diff
@@ -90,14 +90,8 @@ h3(#restclient). RestClient
...
-*** @(RSC19b3)@ @success@ is a boolean attribute which is @true@ when the HTTP status code indicates sucess i.e. @200@ <= @statusCode@ < @300@
-*** @(RSC19b4)@ @errorCode@ is an integer attribute populated with the error code if the @X-Ably-Errorcode@ HTTP header is sent in the response
-*** @(RSC19b5)@ @errorMessage@ is a string attribute populated with the error code if the @X-Ably-Errormessage@ HTTP header is sent in the response
-*** @(RSC19b6)@ @headers@ is an Array of key value pairs for each response header
+** @(RSC19a)@ Method signature is @request(string method, string path, Hash params?, JsonObject body?, Hash headers?) -> HttpPaginatedResponse@ with arguments: @method@ is a valid "HTTP verb":https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html such as @"GET"@; @path@ is the path component of the URL such as @"/channels"@; @params@ and @headers@ are optional arguments containing key value pairs that will include query params and HTTP headers respectively in the request (the argument types can be idiomatic for the language such as @Object@ in the case of Javascript); @body@ is an optional JSON-like object argument that can be easily serialised to MsgPack or JSON
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: Do we support  a `body` argument that serialises to a JSON array?

**Comment by paddybyers** [comment]
I think we should say something about authentication: is this request unconditionally authenticated using the current auth credentials of the library instance?

Other questions I think the documentation should be clear on:

- are any other params supplied by default (eg `clientId` )?

- does the caller need to provide `Accept` and `Content-Type` headers and/or params or does the library do that?

- does the library support multi-valued headers?

- is there any support for binary request or response bodies?

File: content/client-lib-development-guide/features.textile (hunk 1 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: Yes, I should be more explicit on that, JSON object or JSON array

**Comment by mattheworiordan** [comment]
@paddybyers see 23aeaa4 and cbd5e12

I did not mention `clientId` as I am not sure how it is relevant i.e. if `clientId` is implicit then the client library will not send it, and if it's explicit then it would be a param, header or part of the request body.

File: content/client-lib-development-guide/features.textile [hunk 2]
```diff
@@ -326,14 +333,14 @@ h3(#realtime-connection). Connection
...
+** @(RTN14b)@ If a connection request fails due to an @ERROR@ @ProtocolMessage@ being sent to the client containing a token error (@statusCode@ value of 401 and error @code@ value in the range @40140 <= code < 40150@), then if the token is renewable, the error should not be emitted, a single attempt to create a new token will be made and a new connection attempt will be initiated using the newly created token. If the attempt to create a new token fails, or the subsequent connection attempt fails due to another token error, then the connection will transition to the @FAILED@ state and the @Connection#errorReason@ will be set
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: I couldn't work out what has changed in this line

File: content/client-lib-development-guide/features.textile [hunk 3]
```diff
@@ -333,14 +333,14 @@ h3(#realtime-connection). Connection
...
+** @(RTN15h)@ If a @DISCONNECTED@ message is received from Ably, then that transport will subsequently be closed by Ably. If the @DISCONNECTED@ message contains a token error (@statusCode@ value of 401 and error @code@ value in the range @40140 <= code < 40150@), if the token is renewable, the error should not emitted, a single attempt to create a new token will be made and a new connection attempt will be initiated using the new token. If the token creation fails, the next connection attempt fails due to a token error, or the token was not renewable, the connection will transition to the @FAILED@ state and the @Connection#errorReason@ will be set
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: typo "should not emitted" -> "should not be emitted"

**Comment by paddybyers** [comment]
LGTM

File: content/client-lib-development-guide/features.textile (hunk 2 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: @paddybyers see https://github.com/ably/docs/pull/133/commits/175e6ccdfe3f68e595a5eadddb8146264e91764c, it's just a fix to a typo in those error code ranges

**Comment by mattheworiordan** [comment]
@SimonWoolf you happy with this too?  I'd like to merge shortly unless you object

**Comment by mattheworiordan** [comment]
@SimonWoolf you happy with this too?  I'd like to merge shortly unless you object

**Comment by SimonWoolf** [comment]
yep :+1:

**Pull request merged**